### PR TITLE
Add model interface support for remote and local custom models

### PR DESCRIPTION
**Pull request by joshpalis**
### Description
Adds support for the `interface` field of model registration. Currently only adding support for the `register_remote_model` and `register_local_custom_model` since pretrained/ local sparse encoding models do not currently work with the model interface. 

Registering and deploying remote model with model interface : 
```
curl -i -XPOST "localhost:9200/_plugins/_flow_framework/workflow?provision=true" -H "Content-Type:application/json" --data '{"name":"createconnector-registerremotemodel-deploymodel","description":"test case","use_case":"TEST_CASE","version":{"template":"1.0.0","compatibility":["2.12.0","3.0.0"]},"workflows":{"provision":{"nodes":[{"id":"workflow_step_1","type":"create_connector","user_inputs":{"name":"OpenAI Chat Connector","description":"The connector to public OpenAI model service for GPT 3.5","version":"1","protocol":"http","parameters":{"endpoint":"api.openai.com","model":"gpt-3.5-turbo"},"credential":{"openAI_key":"12345"},"actions":[{"action_type":"predict","method":"POST","url":"https://${parameters.endpoint}/v1/chat/completions"}]}},{"id":"workflow_step_2","type":"register_remote_model","previous_node_inputs":{"workflow_step_1":"connector_id"},"user_inputs":{"name":"openAI-gpt-3.5-turbo","description":"test model","interface":{"input":{"properties":{"parameters":{"properties":{"messages":{"type":"string","description":"This is a test description field"}}}}},"output":{"properties":{"inference_results":{"type":"array","items":{"type":"object","properties":{"output":{"type":"array","items":{"properties":{"name":{"type":"string","description":"This is a test description field"},"dataAsMap":{"type":"object","description":"This is a test description field"}}},"description":"This is a test description field"},"status_code":{"type":"integer","description":"This is a test description field"}}},"description":"This is a test description field"}}}},"deploy":true,"node_timeout":"15s"}}]}}}'
HTTP/1.1 201 Created
X-OpenSearch-Version: OpenSearch/3.0.0-SNAPSHOT (opensearch)
content-type: application/json; charset=UTF-8
content-length: 38

{"workflow_id":"gNg7O48BaP6uKfWC1aKh"}

curl -i -XGET "http://localhost:9200/_plugins/_ml/models/hNg7O48BaP6uKfWC2qJv?pretty"
HTTP/1.1 200 OK
X-OpenSearch-Version: OpenSearch/3.0.0-SNAPSHOT (opensearch)
content-type: application/json; charset=UTF-8
content-length: 1318

{
  "name" : "openAI-gpt-3.5-turbo",
  "model_group_id" : "gtg7O48BaP6uKfWC2aJg",
  "algorithm" : "REMOTE",
  "model_version" : "1",
  "description" : "test model",
  "model_state" : "DEPLOYED",
  "created_time" : 1714685729273,
  "last_updated_time" : 1714685729735,
  "last_deployed_time" : 1714685729734,
  "auto_redeploy_retry_times" : 0,
  "planning_worker_node_count" : 1,
  "current_worker_node_count" : 1,
  "planning_worker_nodes" : [
    "TpCVBoolSOCJOR4dkq66ng"
  ],
  "deploy_to_all_nodes" : true,
  "is_hidden" : false,
  "connector_id" : "gdg7O48BaP6uKfWC2KIR",
  "interface" : {
    "output" : "{\"properties\":{\"inference_results\":{\"description\":\"This is a test description field\",\"type\":\"array\",\"items\":{\"type\":\"object\",\"properties\":{\"output\":{\"description\":\"This is a test description field\",\"type\":\"array\",\"items\":{\"properties\":{\"name\":{\"description\":\"This is a test description field\",\"type\":\"string\"},\"dataAsMap\":{\"description\":\"This is a test description field\",\"type\":\"object\"}}}},\"status_code\":{\"description\":\"This is a test description field\",\"type\":\"integer\"}}}}}}",
    "input" : "{\"properties\":{\"parameters\":{\"properties\":{\"messages\":{\"description\":\"This is a test description field\",\"type\":\"string\"}}}}}"
  }
}
```

For registering and deploying local custom model :
```
curl -i -XPOST "localhost:9200/_plugins/_flow_framework/workflow?provision=true" -H "Content-Type:application/json" --data '{"name":"registermodelgroup-registerlocalmodel-deploymodel","description":"test case","use_case":"TEST_CASE","version":{"template":"1.0.0","compatibility":["2.12.0","3.0.0"]},"workflows":{"provision":{"nodes":[{"id":"workflow_step_1","type":"register_local_custom_model","user_inputs":{"node_timeout":"60s","name":"all-MiniLM-L6-v2","version":"1.0.0","description":"test model","model_format":"TORCH_SCRIPT","function_name":"SPARSE_TOKENIZE","model_content_hash_value":"c15f0d2e62d872be5b5bc6c84d2e0f4921541e29fefbef51d59cc10a8ae30e0f","model_type":"bert","embedding_dimension":"384","framework_type":"sentence_transformers","all_config":"{\"_name_or_path\":\"nreimers/MiniLM-L6-H384-uncased\",\"architectures\":[\"BertModel\"],\"attention_probs_dropout_prob\":0.1,\"gradient_checkpointing\":false,\"hidden_act\":\"gelu\",\"hidden_dropout_prob\":0.1,\"hidden_size\":384,\"initializer_range\":0.02,\"intermediate_size\":1536,\"layer_norm_eps\":1e-12,\"max_position_embeddings\":512,\"model_type\":\"bert\",\"num_attention_heads\":12,\"num_hidden_layers\":6,\"pad_token_id\":0,\"position_embedding_type\":\"absolute\",\"transformers_version\":\"4.8.2\",\"type_vocab_size\":2,\"use_cache\":true,\"vocab_size\":30522}","url":"https://artifacts.opensearch.org/models/ml-models/huggingface/sentence-transformers/all-MiniLM-L6-v2/1.0.1/torch_script/sentence-transformers_all-MiniLM-L6-v2-1.0.1-torch_script.zip","interface":{"input":{"properties":{"text_docs":{"type":"array"},"return_number":{"type":"boolean"},"target_response":{"type":"array"}}}}}},{"id":"workflow_step_2","type":"deploy_model","previous_node_inputs":{"workflow_step_1":"model_id"}}],"edges":[{"source":"workflow_step_1","dest":"workflow_step_2"}]}}}'
HTTP/1.1 201 Created
X-OpenSearch-Version: OpenSearch/3.0.0-SNAPSHOT (opensearch)
content-type: application/json; charset=UTF-8
content-length: 38

{"workflow_id":"lch-O48BUXkJrbtALpC3"}

curl -i -XGET "http://localhost:9200/_plugins/_ml/models/mMh-O48BUXkJrbtAMZCk?pretty"
HTTP/1.1 200 OK
X-OpenSearch-Version: OpenSearch/3.0.0-SNAPSHOT (opensearch)
content-type: application/json; charset=UTF-8
content-length: 1687

{
  "name" : "all-MiniLM-L6-v2",
  "model_group_id" : "lsh-O48BUXkJrbtAMJCA",
  "algorithm" : "SPARSE_TOKENIZE",
  "model_version" : "1",
  "description" : "test model",
  "model_format" : "TORCH_SCRIPT",
  "model_state" : "DEPLOYED",
  "model_content_size_in_bytes" : 91790008,
  "model_content_hash_value" : "c15f0d2e62d872be5b5bc6c84d2e0f4921541e29fefbef51d59cc10a8ae30e0f",
  "model_config" : {
    "model_type" : "bert",
    "embedding_dimension" : 384,
    "framework_type" : "SENTENCE_TRANSFORMERS",
    "all_config" : "{\"_name_or_path\":\"nreimers/MiniLM-L6-H384-uncased\",\"architectures\":[\"BertModel\"],\"attention_probs_dropout_prob\":0.1,\"gradient_checkpointing\":false,\"hidden_act\":\"gelu\",\"hidden_dropout_prob\":0.1,\"hidden_size\":384,\"initializer_range\":0.02,\"intermediate_size\":1536,\"layer_norm_eps\":1e-12,\"max_position_embeddings\":512,\"model_type\":\"bert\",\"num_attention_heads\":12,\"num_hidden_layers\":6,\"pad_token_id\":0,\"position_embedding_type\":\"absolute\",\"transformers_version\":\"4.8.2\",\"type_vocab_size\":2,\"use_cache\":true,\"vocab_size\":30522}"
  },
  "created_time" : 1714690076965,
  "last_updated_time" : 1714690083895,
  "last_registered_time" : 1714690081144,
  "last_deployed_time" : 1714690083894,
  "auto_redeploy_retry_times" : 0,
  "total_chunks" : 10,
  "planning_worker_node_count" : 1,
  "current_worker_node_count" : 1,
  "planning_worker_nodes" : [
    "zTtalEMSTRGrYfx29Gh4sQ"
  ],
  "deploy_to_all_nodes" : true,
  "is_hidden" : false,
  "interface" : {
    "input" : "{\"properties\":{\"target_response\":{\"type\":\"array\"},\"return_number\":{\"type\":\"boolean\"},\"text_docs\":{\"type\":\"array\"}}}"
  }
}
```
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -18,6 +18,7 @@ Inspired from [Keep a Changelog](https://keepachangelog.com/en/1.1.0/)
 - Add guardrails to default use case params ([#658](https://github.com/opensearch-project/flow-framework/pull/658))
 - Allow strings for boolean workflow step parameters ([#671](https://github.com/opensearch-project/flow-framework/pull/671))
 - Add optional delay parameter to no-op step ([#674](https://github.com/opensearch-project/flow-framework/pull/674))
+- Add model interface support for remote and local custom models ([#701](https://github.com/opensearch-project/flow-framework/pull/701))
```
Comment (dbwiddis) [inline CHANGELOG.md:21]: also add to 2.14 release notes

File: src/main/java/org/opensearch/flowframework/common/CommonValue.java [hunk 2]
```diff
@@ -172,6 +172,8 @@ private CommonValue() {}
     public static final String GUARDRAILS_FIELD = "guardrails";
     /** Delay field */
     public static final String DELAY_FIELD = "delay";
+    /** Model Interface Field */
+    public static final String MODEL_INTERFACE = "interface";
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/flowframework/common/CommonValue.java:176]: let's just call this INTERFACE_FIELD

File: src/main/java/org/opensearch/flowframework/util/ParseUtils.java [hunk 3]
```diff
@@ -451,4 +451,25 @@ public static String removingBackslashesAndQuotesInArrayInJsonString(String inpu
...
+     */
+    public static Map<String, String> convertStringToObjectMapToStringToStringMap(Map<String, Object> stringToObjectMap) throws Exception {
+        try (Jsonb jsonb = JsonbBuilder.create()) {
+            Map<String, String> stringToStringMap = new HashMap<>();
+            for (Map.Entry<String, Object> entry : stringToObjectMap.entrySet()) {
+                Object value = entry.getValue();
+                if (value instanceof String) {
+                    stringToStringMap.put(entry.getKey(), (String) value);
+                } else {
+                    stringToStringMap.put(entry.getKey(), jsonb.toJson(value));
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/flowframework/util/ParseUtils.java:469]: does this handle nulls? If not we should do a null check.

**Comment by amitgalitz** [comment]
Instead of relaying on customer to escape strings and having to figure out how to give that user input we can use `ParseUtils.parseArbitraryStringToObjectMapToString(configurationsMap);` like we do for create index and pipelines so we do it for them.

**Comment by amitgalitz** [comment]
> Instead of relaying on customer to escape strings and having to figure out how to give that user input we can use `ParseUtils.parseArbitraryStringToObjectMapToString(configurationsMap);` like we do for create index and pipelines so we do it for them.

ignore this comment :)

File: src/main/java/org/opensearch/flowframework/util/ParseUtils.java (hunk 3 above)
Comment (joshpalis) [reply src/main/java/org/opensearch/flowframework/util/ParseUtils.java:469]: if the request body is misconfigured then we return the error : `{"error":"IOException: template content invalid for specified Content-Type."` Tested with an empty `description` value in the interface input

**Review by dbwiddis — APPROVED** [review]

**Review by amitgalitz — APPROVED** [review]

**Pull request merged**